### PR TITLE
Split out logic for skipping tests to re-use in other cases

### PR DESCRIPTION
**Pull request by bryanmr**
Summary:
To change what tests can be run in various cases, the check logic should be moved to functions and variables that can be changed.

One challenge here is that decorators don't have dynamic functionality. If something is read in when imported and then changed afterwards, it will not actually change. This means we need to separate out the variables that need to be changed for our use case.

Those are put into common_distributed.py and can be changed before importing the distributed_test.py code.

The use case is to add new backends to the tests and split it into tests that can be ran on demand as a separate instance. To do so, you would change DistTestSkipCases after importing it into a launcher or a setup script and then load distributed_test.

Test Plan:
See Signals

Differential Revision: D31906947

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### CODE REVIEW

File: torch/testing/_internal/common_distributed.py
Comment (bryanmr) [reply torch/testing/_internal/common_distributed.py]: I went with DistTestCases

File: torch/testing/_internal/distributed/distributed_test.py
Comment (mrshenli) [inline torch/testing/_internal/distributed/distributed_test.py:834]: This is prior to your change. Do you know why do we need both `require_backend` and `require_backends_available` here?

**Review by mrshenli — APPROVED** [review]
code LGTM. I only have some minor comments. Stamped to unblock. Please address before landing.

File: torch/testing/_internal/common_distributed.py
Comment (bryanmr) [reply torch/testing/_internal/common_distributed.py:80]: I had some weirdness with creating an empty set. I agree, None isn't right. I'll poke at it again and see if I can make it clean.

File: torch/testing/_internal/common_distributed.py
Comment (bryanmr) [reply torch/testing/_internal/common_distributed.py:72]: I just got rid of skip_messages, I think that is cleaner and easier to understand when glancing through the code.

File: torch/testing/_internal/distributed/distributed_test.py
Comment (bryanmr) [reply torch/testing/_internal/distributed/distributed_test.py:834]: I don't think we need both. `require_backends_available` is a stricter test and checks to make sure all of the backends in the list are available, which can fail in the case where NCCL isn't available in particular. My OSS setup can't run on my CPU only system because of NCCL not showing up as available, as an example.

There are cases where it would be useful to have a check for backend availability. It doesn't seem to actually be using it as I would expect, though.

https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/distributed/distributed_test.py#L1200

I can submit a separate patch to remove some of these and clean up where both might actually be needed. The question is if we want to check for availability or just to match on the list. I'd say to match on the list AKA `require_backend`.

**Pull request closed without merging**
